### PR TITLE
Addition of q_networks

### PR DESCRIPTION
**Pull request by nishantkr18**
Hi! I would like to suggest some changes in the structure of DQN structure. Instead of passing FFN model directly into agent, we can pass it to a QNetwork, and that in turn will be passed to the agent. This would have two advantages:
1) It will be easier to add duelingDQN; just by adding two extra layers for Value and Advantage functions and making Predict, Forward, Backward and Update functions. We could use the same q_learning for vanillaDQN, DoubleDQN and DuelingDQN.
1) We may pass a default FFN as model in agent for DQN.
The tests would then change from 
```
  FFN<MeanSquaredError<>, GaussianInitialization> model(MeanSquaredError<>(),
      GaussianInitialization(0, 0.001));
  model.Add<Linear<>>(4, 128);
  model.Add<ReLULayer<>>();
  model.Add<Linear<>>(128, 128);
  model.Add<ReLULayer<>>();
  model.Add<Linear<>>(128, 2);

  // Set up the policy and replay method.
  GreedyPolicy<CartPole> policy(1.0, 1000, 0.1, 0.99);
  RandomReplay<CartPole> replayMethod(10, 10000);

  TrainingConfig config;
  config.StepSize() = 0.01;
  config.Discount() = 0.9;
  config.TargetNetworkSyncInterval() = 100;
  config.ExplorationSteps() = 100;
  config.DoubleQLearning() = false;
  config.StepLimit() = 200;

  // Set up DQN agent.
  QLearning<CartPole, decltype(model), AdamUpdate, decltype(policy)>
      agent(std::move(config), std::move(model), std::move(policy),
      std::move(replayMethod));
```
**to**
```
    // Set up the policy and replay method.
    GreedyPolicy<CartPole> policy(1.0, 1000, 0.1, 0.99);
    RandomReplay<CartPole> replayMethod(10, 10000);
    QNetwork<> network();

    TrainingConfig config;
    config.StepSize() = 0.01;
    config.Discount() = 0.9;
    config.TargetNetworkSyncInterval() = 100;
    config.ExplorationSteps() = 100;
    config.DoubleQLearning() = false;
    config.StepLimit() = 200;

    // Set up DQN agent.
    QLearning<CartPole, decltype(network), AdamUpdate, decltype(policy)>
        agent(std::move(config), std::move(network), std::move(policy),
              std::move(replayMethod));
```
I have added the Qnetwork file without proper documentation and completion as of now, just to get some reviews and suggestions..

### CODE REVIEW

File: src/mlpack/methods/reinforcement_learning/q_network.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+namespace mlpack {
+namespace rl {
+
+template <typename FeedForwardNetworkType>
+class QNetwork
+{
+ public:
+
```
Comment (birm) [inline src/mlpack/methods/reinforcement_learning/q_network.hpp]:
```suggestion
```

File: src/mlpack/methods/reinforcement_learning/q_network.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  QNetwork()
+  {}
+
```
Comment (birm) [inline src/mlpack/methods/reinforcement_learning/q_network.hpp]:
```suggestion

```

File: src/mlpack/methods/reinforcement_learning/q_network.hpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+class QNetwork
+{
+ public:
+
+  QNetwork()
+  {}
+ 
+  QNetwork(FeedForwardNetworkType network) : network(std::move(network))
+  { /* Nothing to do here. */ }
+
```
Comment (birm) [inline src/mlpack/methods/reinforcement_learning/q_network.hpp]:
```suggestion

```

File: src/mlpack/methods/reinforcement_learning/q_network.hpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
+/**
+ * @file q_network.hpp
+ * @author Nishant Kumar
+ *
+ * This file is consists of the q network.
```
Comment (bisakhmondal) [inline src/mlpack/methods/reinforcement_learning/q_network.hpp]:
```suggestion
 * This file consists of the q network.
```

Comment (kartikdutt18) [reply src/mlpack/methods/reinforcement_learning/q_network.hpp]: We generally use this statement, Definition or Implementation of Q-Network. Thanks 👍

File: src/mlpack/methods/reinforcement_learning/q_network.hpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+#include <mlpack/prereqs.hpp>
+#include "training_config.hpp"
+
+
+namespace mlpack {
+namespace rl {
+
+template <typename FeedForwardNetworkType>
+class QNetwork
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_network.hpp]: Form the example it looks like the `QNetwork` has a predefined layer structure?

Comment (nishantkr18) [reply src/mlpack/methods/reinforcement_learning/q_network.hpp]: Yeah, that's what I meant.. although, is 
```
template <typename FeedForwardNetworkType = FFN<MeanSquaredError<>, GaussianInitialization>>
class QNetwork
``` 
valid? Sorry if it seems like a stupid question..

File: src/mlpack/methods/reinforcement_learning/q_network.hpp (hunk 4 above)
Comment (nishantkr18) [reply src/mlpack/methods/reinforcement_learning/q_network.hpp]: I would do the changes with the upcoming commit. Thanks. :+1:

File: src/mlpack/methods/reinforcement_learning/q_network.hpp (hunk 5 above)
Comment (zoq) [reply src/mlpack/methods/reinforcement_learning/q_network.hpp]: Yes, looks valid to me, does this result in any error?

**Comment by kartikdutt18** [comment]
Hey @nishantkr18, The macOS failure seems unrelated. I think if you rebased it should be fixed. Thanks a lot. 👍

File: src/mlpack/methods/reinforcement_learning/q_network.hpp (hunk 5 above)
Comment (nishantkr18) [reply src/mlpack/methods/reinforcement_learning/q_network.hpp]:
```
error: ‘FFN’ does not name a type
 template <typename FeedForwardNetworkType = FFN<MeanSquaredError<>, GaussianInitialization>>
```
Thats the error I get

**Comment by nishantkr18** [comment]
I have created a new folder for q_networks which currently contains the simple DQN by the name `vanillaDQN`. I'll add the other types of Q networks in a separate PR?

File: src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+
+namespace mlpack {
+namespace rl {
+
+using namespace mlpack::ann;
+
+template <typename NetworkType = FFN<MeanSquaredError<>,
+                                    GaussianInitialization>>
+class VanillaDQN
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp:28]: Do you mind to go with another name, I think most people will associate the name with the network from the paper, which isn't correct. Maybe `SimpleDQN` or `MinimalDQN` work as well, maybe you have another name in mind?

File: src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+
+namespace mlpack {
+namespace rl {
+
+using namespace mlpack::ann;
+
+template <typename NetworkType = FFN<MeanSquaredError<>,
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp:26]: Can you add a template parameter description for `NetworkType`?

File: src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  VanillaDQN() : network()
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of VanillaDQN class.
+   *
+   * @param inputDim number of inputs.
+   * @param hiddenDim1 number of neurons in hiddenlayer-1.
+   * @param hiddenDim2 number of neurons in hiddenlayer-2.
+   * @param outputDim number of neurons in output layer.
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp:43]:
```suggestion
   * @param inputDim number of inputs.
   * @param hiddenDim1 Number of neurons in hiddenlayer-1.
   * @param hiddenDim2 Number of neurons in hiddenlayer-2.
   * @param outputDim Number of neurons in output layer.
```

Also, what do you think about using `h1, h2` as parameter name?

File: src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+   * output layer function.
+   *
+   * If you want to pass in a parameter and discard the original parameter
+   * object, be sure to use std::move to avoid unnecessary copy.
+   *
+   * @param state Input state.
+   * @param actionValue Matrix to put output action values of states input.
+   */
+  void Predict(const arma::mat state,
+             arma::mat& actionValue)
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp:75]:
```suggestion
  void Predict(const arma::mat state, arma::mat& actionValue)
```

File: src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  /**
+   * Perform the forward pass of the states in real batch mode.
+   *
+   * @param state The input state.
+   * @param target The predicted target.
+   */
+  void Forward(const arma::mat state,
+             arma::mat& target)
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp:87]:
```suggestion
  void Forward(const arma::mat state, arma::mat& target)
```

File: src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  /**
+   * Perform the backward pass of the state in real batch mode.
+   *
+   * @param state The input state.
+   * @param target The training target.
+   * @return gradient The gradient.
+   */
+  void Backward(const arma::mat state,
+             arma::mat& target,
+             arma::mat& gradient)
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp:115]:
```suggestion
  void Backward(const arma::mat state, arma::mat& target, arma::mat& gradient)
```

File: src/mlpack/tests/reward_clipping_test.cpp [hunk 12]
```diff
@@ -63,13 +64,7 @@ BOOST_AUTO_TEST_CASE(RewardClippedAcrobotWithDQN)
...
   {
     // Set up the network.
-    FFN<MeanSquaredError<>, GaussianInitialization> model(MeanSquaredError<>(),
-        GaussianInitialization(0, 0.001));
-    model.Add<Linear<>>(4, 64);
-    model.Add<ReLULayer<>>();
-    model.Add<Linear<>>(64, 32);
-    model.Add<ReLULayer<>>();
-    model.Add<Linear<>>(32, 3);
+    VanillaDQN<> model(4, 64, 32, 3);
```
Comment (zoq) [inline src/mlpack/tests/reward_clipping_test.cpp:67]: Do you think we should show in a single test, it's also possible to manually specifiy the network, completely without using `VanillaDQN`?

File: src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp (hunk 6 above)
Comment (nishantkr18) [reply src/mlpack/methods/reinforcement_learning/q_networks/vanilla_dqn.hpp:28]: Yeah, I guess `SimpleDQN` should do.. Sounds more relatable

File: src/mlpack/tests/reward_clipping_test.cpp (hunk 12 above)
Comment (nishantkr18) [reply src/mlpack/tests/reward_clipping_test.cpp:67]: Hmm. I think u mean this:
```
    FFN<MeanSquaredError<>, GaussianInitialization>
        network(MeanSquaredError<>(),
                GaussianInitialization(0, 0.001));
    network.Add<Linear<>>(6, 256);
    network.Add<ReLULayer<>>();
    network.Add<Linear<>>(256, 3);

    // Create custom network type
    VanillaDQN<decltype(network)> model(std::move(network));
```
This is taken from the test for `DoublePoleCartWithDQN`, showing how we can manually specify the network. But I could add a separate test for CartPole with DQN as well, showing how to manually specify.?

I'm afraid we wont be able to directly pass the network to `QLearning`(ie without using `VanillaDQN`), as I've used the method `ResetParametersIfEmpty()` in QLearning which is not present in FNN. Would that be fine?

Comment (zoq) [reply src/mlpack/tests/reward_clipping_test.cpp]: We could rename `ResetParametersIfEmpty` to `ResetParameters` and revert the change, that way we can still pass a vanilla network, maybe I missed something? There is already a test that uses the copy constructor, but would be great if we could provide backward compatibility.

Comment (nishantkr18) [reply src/mlpack/tests/reward_clipping_test.cpp]: Yeah, you are right.. Actually I didn't notice that `Parameters()` had already been added to `SimpleDQN`.. 
Anyways, now I've done the necessary changes.. Kindly have a look..

**Comment by sriramsk1999** [comment]
Hi @nishantkr18 , I was having a look at the code and I had a few questions. Wouldn't structuring the network this way be too restrictive? As in a `simpleDQN` would always be a two-layer network with `Linear` and `ReLu`? 

What would the process be if I wanted to use a DQN of a different architecture, say 3-layer `Convolution` with `LeakyReLU`? I hope my question is clear :)

**Comment by nishantkr18** [comment]
> Hi @nishantkr18 , I was having a look at the code and I had a few questions. Wouldn't structuring the network this way be too restrictive? As in a `simpleDQN` would always be a two-layer network with `Linear` and `ReLu`?
> 
> What would the process be if I wanted to use a DQN of a different architecture, say 3-layer `Convolution` with `LeakyReLU`? I hope my question is clear :)

Hi @sriramsk1999 ! One can easily pass custom network architectures directly into `QLearning` or via `SimpleDQN`, as is done [here](https://github.com/mlpack/mlpack/pull/2317/files#diff-ae22bcf6dd688ff7cac3d0e280fe97d5R378). 
The purpose of `q_networks` is that when other extensions are added to DQN, whose network structures are different from each other, instead of creating the entire network structure in the tests itself, one can easily use the preexisting networks from `q_networks`. But again, custom network support is still available.. 
I hope that makes it clear. Let me know if there are any other questions :)

**Comment by sriramsk1999** [comment]
Ah okay, I think I understand now. I thought that `q_networks` was supposed to replace the existing way to add a network to the `QLearning` agent.

If I understood you correctly, it is supplementary to the existing method and acts as a shortcut when using the commonly used architectures. Thanks for the clarification and nice work. :)

**Pull request merged**